### PR TITLE
Correct `generateName` doc when in conflict

### PR DESCRIPTION
**Pull request by ninjaprox**
The doc seems to be outdated. The [code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apimachinery/pkg/api/errors/errors.go#L178) actually returns `409` in case the generated name already exists. This PR is to update the doc to avoid confusion.

**Which issue(s) this PR fixes**:

Fixes #

### CODE REVIEW

File: contributors/devel/sig-architecture/api-conventions.md [hunk 1]
```diff
@@ -730,11 +730,9 @@ passed). The value of this field will be combined with a unique suffix on the
...
 within the rules for Name, and may be truncated by the length of the suffix
 required to make the value unique on the server. If this field is specified, and
-Name is not present, the server will NOT return a 409 if the generated name
-exists - instead, it will either return 201 Created or 504 with Reason
-`ServerTimeout` indicating a unique name could not be found in the time
-allotted, and the client should retry (optionally after the time indicated in
-the Retry-After header).
+Name is not present, the server will return a 409 with Reason `AlreadyExists` if
+the generated name exists, and the client should retry (optionally after the time
+indicated in the Retry-After header).
```
Comment (lavalamp) [inline contributors/devel/sig-architecture/api-conventions.md:735]:
```suggestion
the generated name exists, and the client should retry (after waiting at least the amount of time
indicated in the Retry-After header, if it is present).
```

(I was debating about making this comment or not)

**Comment by lavalamp** [comment]
I think I'd prefer to imply that the header is optional, than to imply that it's optional to obey the header.

**Comment by thockin** [comment]
The current doc implies that apiserver will retry on name-conflict, but it does not seem to actually do so.  See staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go `Store.Create()` which calls `objectMeta.SetName(e.CreateStrategy.GenerateName(...))` one time, before passing off to `e.Storage.Create()` and then looking at the result with `rest.CheckGeneratedNameError()`

I thought maybr this is retried higher up-stack, but I couldn't find it (but this subsystem is not my expertise).  

@lavalamp or @deads2k or delegate thereof - can you verify?

**Comment by liggitt** [comment]
I don't see any facility for server-internal retry on create conflict... all the internal retries I know of are around update

**Comment by thockin** [comment]
That seems unfortunate - minor real impact but surprising that a server-side allocation would fail with a conflict.

**Comment by lavalamp** [comment]
I think it may have originally retried for you a few times, but that behavior seems to be gone.

The concerning part is actually what happens when a collection gets particularly full -- it becomes exponentially unlikely that a random name is free. That could be a lot of retries!

**Comment by thockin** [comment]
Is this a bug, in your opinion?

**Comment by lavalamp** [comment]
I think it's very surprising behavior, at a minimum. Maybe we should change it, but in the meantime I do think we should fix the documentation.

**Comment by thockin** [comment]
What I really meant was - should I or someone file a k/k bug?

**Comment by lavalamp** [comment]
Sure, it's probably at least worth a canonical discussion.

File: contributors/devel/sig-architecture/api-conventions.md (hunk 1 above)
Comment (ninjaprox) [reply contributors/devel/sig-architecture/api-conventions.md:735]: I wonder what is the effect of Retry-After header here? I see it in some other errors like server timeout or too many request, which makes sense and understandable. Is there any differences between a client chooses to retry with vs without respecting the header?

**Comment by ninjaprox** [comment]
For more context, I encountered this issue when using Argo, and in one rare case, its workflow name collided. The client has retry mechanism on 500 and from the docs I expected it should have covered the name collision as well. After checking the code, I found out it's inconsistent with the docs.

File: contributors/devel/sig-architecture/api-conventions.md (hunk 1 above)
Comment (lavalamp) [reply contributors/devel/sig-architecture/api-conventions.md:735]: The purpose is to keep the server from being overwhelmed by clients retrying in a hot loop, should the condition take some time to resolve.

**Comment by dims** [comment]
@lavalamp so guidance is to close this?

@ninjaprox could you log a bug instead as @thockin mentioned?

**Comment by ninjaprox** [comment]
@dims I think we should fix the document first as @lavalamp mentioned. https://github.com/kubernetes/kubernetes/issues/115489 is to discuss whether it's a bug or not.

File: contributors/devel/sig-architecture/api-conventions.md (hunk 1 above)
Comment (thockin) [reply contributors/devel/sig-architecture/api-conventions.md:735]: I'm OK to merge this with @lavalamp changes

**Comment by johnbelamaric** [comment]
I don't find this text very clear. It says: "A non-empty value for the field indicates the name will be
made unique (and the name returned to the client will be different than the name
passed). The value of this field will be combined with a unique suffix on the
server if the Name field has not been provided."

then it says

" the server will return a 409 with Reason `AlreadyExists` if
the generated name exists"

which is a direct contradiction of "will be made unique". Yes, after a couple reads it becomes clear - oh, the server won't make it unique, it will try one time to generate a random suffix and fail if that's not unique.

**Comment by johnbelamaric** [comment]
This would be better:

```
A non-empty value for the field indicates the server should attempt to
make the name unique (and the name returned to the client will be different than the name
passed). The value of this field will be combined with a random suffix on the
server if the Name field has not been provided. The provided value must be valid
within the rules for Name, and may be truncated by the length of the suffix. If this field is
specified, and Name is not present, the server will return a 409 with Reason `AlreadyExists` if
the generated name exists, and the client should retry (after waiting at least the amount of time
indicated in the Retry-After header, if it is present).
```

**Pull request merged**